packages/guides/src/Compiler/Passes/ImplicitHyperlinkTargetPass.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+namespace phpDocumentor\Guides\Compiler\Passes;
+
+use phpDocumentor\Guides\Compiler\CompilerPass;
+use phpDocumentor\Guides\Nodes\AnchorNode;
+use phpDocumentor\Guides\Nodes\DocumentNode;
+use phpDocumentor\Guides\Nodes\Node;
+use phpDocumentor\Guides\Nodes\SectionNode;
+
+/**
+ * Resolves the hyperlink target for each section in the document.
+ *
+ * This follows the reStructuredText rules as outlined in:
+ * https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#implicit-hyperlink-targets
+ */
+class ImplicitHyperlinkTargetPass implements CompilerPass
+{
+    public function getPriority(): int
+    {
+        return 20000; // must be run *before* MetasPass
+    }
+
+    public function run(array $documents): array
+    {
+        return array_map(function (DocumentNode $document) {
+            // implicit references must not conflict with explicit ones
+            $knownReferences = $explicitReferences = $this->fetchExplicitReferences($document);
+
+            $nodes = $document->getNodes();
+            $node = current($nodes);
+            do {
+                if ($node instanceof AnchorNode) {
+                    // override implicit section reference if an anchor precedes the section
+                    $key = key($nodes);
+                    $section = next($nodes);
+                    if (!$section instanceof SectionNode) {
+                        prev($nodes);
+                        continue;
+                    }
+
+                    $section->getTitle()->setId($node->getValueString());
+
+                    $document = $document->removeNode($key);
+
+                    continue;
+                }
+
+                if ($node instanceof SectionNode) {
+                    $realId = $sectionId = $node->getTitle()->getId();
```

Comment:
Should this not be part of the parser? Titles are passed around as reference to sections. Changing them would result into possible bugs? As you do this per document, it feels like it could be done with some parser context.

Reply:
Ah, I wasn't aware that we used titles in that way currently.

I would say that we should use the new Metas as the only source-of-truth for references. I think we should somehow pass around a "section id" for references and resolve it using the Metas to a real link in a node transformer (?).

This is hard (and probably not wanted) to implement in a parser, as it would require knowing all explicit references and titles in the document. See e.g. these 2 examples:

```rst
.. hyperlink target: some-title

Some Title
----------

...

.. hyperlink target: some-title-1 (resolve conflict with already known titles)

Some Title
----------
```

```rst
.. hyperlink target: some-title-1 (resolve conflict with explicit hyperlink target below)

Some Title
----------

...

.. _some-title:

Example text.
```

Reply:
You are updating the title id, I guess we need to have IDs equal everywhere. So maybe it's not an issue? 

If I get it right, the order matters, so links will break anyway when a new anchor/title is added